ocaml/xapi/extauth_plugin_ADpbis.ml

```diff
@@ -734,83 +734,94 @@ module AuthADlw : Auth_signature.AUTH_MODULE = struct
-          (* OK, we found this username! *)
-          debug
-            "Request %i/%i to external authentication server successful: user \
-             %s was found"
-            i max username ;
-          (* (2) CA-25427: test (1) above may succeed (because of pbis caching stale AD information) *)
-          (* even though the AD domain is offline (error 32888), usually because /etc/resolv.conf is not *)
-          (* pointing to the AD server. This test should catch if the domain is offline by calling find-by-sid *)
-          (* using a domain SID. We must use a _domain_ SID. A universal SID like S-1-1-0 doesn't work for this test. *)
-          let (_ : (string * string) list) = query_subject_information sid in
-          (* use KRBTGT's domain SID *)
-          debug
-            "Request %i/%i to external authentication server successful: sid \
-             %s was found"
-            i max sid ;
-          true
+  let _is_pbis_server_available max_tries =
+    (* we _need_ to use a username contained in our domain, otherwise the following tests won't work.
+       Microsoft KB/Q243330 article provides the KRBTGT account as a well-known built-in SID in AD
+       Microsoft KB/Q229909 article says that KRBTGT account cannot be renamed or enabled, making
+       it the perfect target for such a test using a username (Administrator account can be renamed) *)
+    let krbtgt = "KRBTGT" in
+    let try_clear_cache () =
+      match get_full_subject_name krbtgt with
+      | exception e ->
+          info "_is_pbis_server_available: failed to get full subject name" ;
+          false
+      | full_username -> (
+        match
+          ignore
+            (pbis_common "/opt/pbis/bin/ad-cache"
+               ["--delete-user"; "--name"; full_username])
         with
-        | Not_found ->
-            (* that means that pbis is responding to at least cached subject queries. *)
-            (* in this case, KRBTGT wasn't found in the AD domain. this usually indicates that the *)
-            (* AD domain is offline/inaccessible to pbis, which will cause problems, specially *)
-            (* to the ssh python hook-script, so we need to try again until KRBTGT is found, indicating *)
-            (* that the domain is online and accessible to pbis queries *)
-            debug
-              "Request %i/%i to external authentication server returned KRBTGT \
-               Not_found, waiting 5 secs to try again"
-              i max ;
-            Thread.delay 5.0 ;
-            (*wait 5 seconds*)
-            (* try again *)
-            test (i + 1)
-        | e ->
-            (* ERROR: anything else means that the server is NOT responding adequately *)
-            debug
-              "Request %i/%i to external authentication server failed, waiting \
-               5 secs to try again: %s"
-              i max
+        | () ->
+            true
+        | exception Not_found ->
```

Comment:
Could you add a comment why this means the service is available?

Reply:
I've added a comment

The purpose of `try_clear_cache` is simply to remove the KRBTGT user from the cache to set up for the 'real test', `try_fetch_sid`. If the cache is populated, then `try_fetch_sid` will not perform an end to end query to the AD server